### PR TITLE
rtl8852au: init at 70bdde2

### PR DESCRIPTION
**Pull request by lonyelon**
## Description of changes

Add the RTL8852AU driver for some devices that require it, like my WiFi antenna. Source at https://github.com/lwfinger/rtl8852au.

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by rjpcasalino** [comment]
this works well. I modified it to fetch the `8852bu` and that worked too! many thanks.

File: pkgs/os-specific/linux/rtl8852au/default.nix
Comment (rjpcasalino) [inline pkgs/os-specific/linux/rtl8852au/default.nix:5]: the date here might need to be updated?

**Review by rjpcasalino — APPROVED** [review]
I think this can be merged? I wonder if the author could add `8852bu` as well since like I've said I've copied this for such a purpose and saw success locally

**Review by Aleksanaa — COMMENTED** [review]
You have to add `rtl8821au = callPackage ../os-specific/linux/rtl8821au { };` in `pkgs/top-level/linux-kernels.nix`, otherwise it won't get added (unlike what `by-name` does)

**Comment by rjpcasalino** [comment]
this probably is a matter of compiling the driver correctly but I get `SAE unsupported: rtl8852bu needs CMD_EXTERNAL_AUTH for SAE` when I attempt to connect to a hidden network with `iwctl`

File: pkgs/os-specific/linux/rtl8852au/default.nix
Comment (lonyelon) [reply pkgs/os-specific/linux/rtl8852au/default.nix:5]: On it!

File: pkgs/os-specific/linux/rtl8852au/default.nix
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/rtl8852au/default.nix:17]:
```suggestion
  postPatch = ''
```

**Comment by lonyelon** [comment]
All changes done! sorry for the long wait, I was off the internet!

Ill look into adding the 8852bu next, as many of you confirm it works.

File: pkgs/os-specific/linux/rtl8852au/default.nix
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/rtl8852au/default.nix:51]: Please add yourself as a maintainer by adding an alphabetically sorted entry to `maintainers/maintainer-list.nix` and commit this change into a separate commit with the message ``maintainers: add NAME`` into this PR.

Comment (lonyelon) [reply pkgs/os-specific/linux/rtl8852au/default.nix:51]: Done!

**Comment by kirillrdy** [comment]
@lonyelon can you change order of commits. otherwise first commit is referencing an attribute that doesn't exist

**Comment by lonyelon** [comment]
> @lonyelon can you change order of commits. otherwise first commit is referencing an attribute that doesn't exist

Done! <3

I had not even thought about that lol

File: pkgs/os-specific/linux/rtl8852au/default.nix
Comment (kirillrdy) [inline pkgs/os-specific/linux/rtl8852au/default.nix:19]: `--replace` is deprecated,

https://nixos.org/manual/nixpkgs/unstable/#fun-substitute-replace-fail

Comment (lonyelon) [reply pkgs/os-specific/linux/rtl8852au/default.nix:19]: What do I use?

**Comment by lonyelon** [comment]
I changed every '--replace' with '--replace-fail' as asked by @kirillrdy. I also detected an unnecessary replace thanks to it.
All is fixed now.

**Comment by rjpcasalino** [comment]
@lonyelon yes! Would love to test. I was seeing strange issues I'm trying to see if they are related to `8852bu` but haven't isolated it yet so would be happy to further test.

**Review by kirillrdy — COMMENTED** [review]
Result of `nixpkgs-review pr 279702` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>12 packages failed to build:</summary>
  <ul>
    <li>linuxKernel.packages.linux_4_19.rtl8852au</li>
    <li>linuxPackages_4_19_hardened.rtl8852au (linuxKernel.packages.linux_4_19_hardened.rtl8852au)</li>
    <li>linuxPackages_5_10_hardened.rtl8852au (linuxKernel.packages.linux_5_10_hardened.rtl8852au)</li>
    <li>linuxPackages_5_15_hardened.rtl8852au (linuxKernel.packages.linux_5_15_hardened.rtl8852au)</li>
    <li>linuxPackages_5_4_hardened.rtl8852au (linuxKernel.packages.linux_5_4_hardened.rtl8852au)</li>
    <li>linuxPackages_6_1_hardened.rtl8852au (linuxKernel.packages.linux_6_1_hardened.rtl8852au)</li>
    <li>linuxPackages_hardened.rtl8852au (linuxPackages_6_6_hardened.rtl8852au)</li>
    <li>linuxPackages_6_7_hardened.rtl8852au (linuxKernel.packages.linux_6_7_hardened.rtl8852au)</li>
    <li>linuxPackages_latest.rtl8852au (linuxKernel.packages.linux_6_8.rtl8852au)</li>
    <li>linuxPackages_latest-libre.rtl8852au (linuxKernel.packages.linux_latest_libre.rtl8852au)</li>
    <li>linuxPackages_lqx.rtl8852au (linuxKernel.packages.linux_lqx.rtl8852au)</li>
    <li>linuxPackages_zen.rtl8852au (linuxKernel.packages.linux_zen.rtl8852au)</li>
  </ul>
</details>
<details>
  <summary>9 packages built:</summary>
  <ul>
    <li>linuxKernel.packages.linux_5_10.rtl8852au</li>
    <li>linuxKernel.packages.linux_5_15.rtl8852au</li>
    <li>linuxKernel.packages.linux_5_4.rtl8852au</li>
    <li>linuxKernel.packages.linux_6_1.rtl8852au</li>
    <li>linuxPackages.rtl8852au (linuxKernel.packages.linux_6_6.rtl8852au)</li>
    <li>linuxKernel.packages.linux_6_7.rtl8852au</li>
    <li>linuxPackages-libre.rtl8852au (linuxKernel.packages.linux_libre.rtl8852au)</li>
    <li>linuxPackages_xanmod.rtl8852au (linuxKernel.packages.linux_xanmod.rtl8852au)</li>
    <li>linuxPackages_xanmod_latest.rtl8852au (linuxKernel.packages.linux_xanmod_latest.rtl8852au ,linuxPackages_xanmod_stable.rtl8852au)</li>
  </ul>
</details>

**Comment by lonyelon** [comment]
I can test the zen, libre and hardened kernels no problem. I have an old thinkpad where I can plug the antenna and check. It just feels weird that it fails on those.

**Comment by rjpcasalino** [comment]
any luck with further testing? I found out my motherboard was dying so that was my issue but I've got a new one now so I can go ahead and help out testing

**Comment by kirillrdy** [comment]
@K900 asking for review, as I don't have much experience in the nixpkgs kernel area

otherwise, I don't see reasons not to merge it

**Comment by lonyelon** [comment]
> Do we really have to? This hardware seems to be supported by rtw89 in latest kernels.

Latest as in 6.8? In 6.6 I've had no luck.

**Comment by K900** [comment]
Please try 6.8.

**Comment by rjpcasalino** [comment]
is 8852bu also supported in 6.8? I am guessing yes and if so I think we can close this but many thanks @lonyelon regardless

**Comment by rjpcasalino** [comment]
tried 6.8 and testing but doesn't seem to see the usb adapter anymore but perhaps I have a misconfiguration somewhere

**Comment by K900** [comment]
Oh I see, the U variants are USB, which is not supported in mainline yet. I guess we have to deal with this for a bit longer then...

**Comment by lonyelon** [comment]
Thanks everyone. I'll now start working on the RTL8852BU PR here: https://github.com/NixOS/nixpkgs/pull/306247.

**Pull request merged**
